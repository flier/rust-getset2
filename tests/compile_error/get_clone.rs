use getset2::Getter;

#[derive(Getter)]
pub struct Struct {
    #[get(clone)] // #[get(clone)] is not allowed for a field that is not implements `Clone`
    field: Foobar,
}

#[derive(Getter)]
#[get(clone)] // #[get(clone)] is not allowed for a field that is not implements `Clone`
pub struct Struct2 {
    field: Foobar,
}

pub struct Foobar;

fn main() {}
