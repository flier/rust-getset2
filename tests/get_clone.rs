use std::path::{Path, PathBuf};

use getset2::Getter;

#[derive(Getter)]
struct Foo<'a> {
    #[get(clone)]
    pub clone_field: PathBuf,

    #[get(clone)]
    pub clone_ref_field: &'a PathBuf,
}

#[test]
fn get_clone() {
    let p = PathBuf::from("/tmp");
    let foo = Foo {
        clone_field: PathBuf::from("/tmp"),
        clone_ref_field: &p,
    };

    assert_eq!(foo.clone_field(), PathBuf::from("/tmp"));
    assert_eq!(foo.clone_ref_field(), Path::new("/tmp"));
}
