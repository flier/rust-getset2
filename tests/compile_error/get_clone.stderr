error[E0277]: the trait bound `Foobar: Clone` is not satisfied
 --> tests/compile_error/get_clone.rs:3:10
  |
3 | #[derive(Getter)]
  |          ^^^^^^ the trait `Clone` is not implemented for `Foobar`
4 | pub struct Struct {
5 |     #[get(clone)] // #[get(clone)] is not allowed for a field that is not implements `Clone`
  |     - required by a bound introduced by this call
  |
help: consider annotating `Foobar` with `#[derive(Clone)]`
  |
15+ #[derive(Clone)]
16| pub struct Foobar;
  |

error[E0277]: the trait bound `Foobar: Clone` is not satisfied
  --> tests/compile_error/get_clone.rs:9:10
   |
9  | #[derive(Getter)]
   |          ^^^^^^ the trait `Clone` is not implemented for `Foobar`
10 | #[get(clone)] // #[get(clone)] is not allowed for a field that is not implements `Clone`
   | - required by a bound introduced by this call
   |
help: consider annotating `Foobar` with `#[derive(Clone)]`
   |
15 + #[derive(Clone)]
16 | pub struct Foobar;
   |
